pkg/runtime/interfaces.go

```diff
@@ -23,64 +23,74 @@ import (
 	"k8s.io/kubernetes/pkg/api/unversioned"
 )
 
-// Codec defines methods for serializing and deserializing API objects.
-type Codec interface {
-	Decoder
-	Encoder
-}
+const (
+	APIVersionInternal    = "__internal"
+	APIVersionUnversioned = "__unversioned"
+)
 
-// Decoder defines methods for deserializing API objects into a given type
-type Decoder interface {
-	Decode(data []byte) (Object, error)
-	// TODO: Remove this method?
-	DecodeToVersion(data []byte, groupVersion unversioned.GroupVersion) (Object, error)
-	DecodeInto(data []byte, obj Object) error
-	// TODO: Remove this method?
-	DecodeIntoWithSpecifiedVersionKind(data []byte, obj Object, groupVersionKind unversioned.GroupVersionKind) error
-
-	DecodeParametersInto(parameters url.Values, obj Object) error
+// Typer retrieves information about an object's group, version, and kind.
+type Typer interface {
+	// ObjectKind returns the version and kind of the provided object, or an
+	// error if the object is not recognized (IsNotRegisteredError will return true).
+	// It returns whether the object is considered unversioned at the same time.
+	ObjectKind(Object) (*unversioned.GroupVersionKind, bool, error)
 }
 
-// Encoder defines methods for serializing API objects into bytes
 type Encoder interface {
-	Encode(obj Object) (data []byte, err error)
-	EncodeToStream(obj Object, stream io.Writer) error
+	// EncodeToStream writes an object to a stream. Override versions may be provided for each group
+	// that enforce a certain versioning. Implementations may return errors if the versions are incompatible,
+	// or if no conversion is defined.
+	EncodeToStream(obj Object, stream io.Writer, overrides ...unversioned.GroupVersion) error
+}
 
-	// TODO: Add method for processing url parameters.
-	// EncodeParameters(obj Object) (url.Values, error)
+type Decoder interface {
+	// Decode attempts to deserialize the provided data using either the innate typing of the scheme or the
+	// default kind, group, and version provided. It returns a decoded object as well as the actual kind, group, and
+	// version decoded, or an error. If into is non-nil, it will be used as the target type and implementations
+	// may choose to use it rather than reallocating an object. However, the object is not guaranteed to be
+	// populated. The returned object type is not guaranteed to match into.
+	Decode(data []byte, defaults *unversioned.GroupVersionKind, into Object) (Object, *unversioned.GroupVersionKind, error)
```

Comment:
Yes.  I'm still playing with the shape of this method, but it returns what
the deserialized value was (I still need to decide whether it's pre or post
defaulting, I started with post).

On Dec 11, 2015, at 6:01 PM, Chao Xu <notifications@github.com> wrote:


>
> -	// TODO: Add method for processing url parameters.
> -	// EncodeParameters(obj Object) (url.Values, error)
> +type Decoder interface {
> +	// Decode attempts to deserialize the provided data using either the innate typing of the scheme or the
> +	// default kind, group, and version provided. It returns a decoded object as well as the actual kind, group, and
> +	// version decoded, or an error. If into is non-nil, it will be used as the target type and implementations
> +	// may choose to use it rather than reallocating an object. However, the object is not guaranteed to be
> +	// populated. The returned object type is not guaranteed to match into.
> +	Decode(data []byte, defaults *unversioned.GroupVersionKind, into Object) (Object, *unversioned.GroupVersionKind, error)

What is innate typing of the scheme?
Does actual kind, group, and version decoded mean the serialized
kind/group/version in data?